Plugins/MaxLevel/MaxLevel.cpp

```diff
@@ -329,43 +329,41 @@ static void LoadSpellKnownTableHook(CNWClass* pClass, CExoString sTable)
     }
 }
 
-// If the player is at level 40 or lower we get the spell known for that class from the normal array,
+// If the player is at class level 40 or lower we get the spell known for that class from the normal array,
 // otherwise we use our map
-static uint8_t GetSpellsKnownPerLevelHook(CNWClass *pClass, uint8_t nLevel, uint8_t nSpellLevel, uint8_t nClass,
-                                             uint16_t nRace, uint8_t nCHABase)
+static uint8_t GetSpellsKnownPerLevelHook(CNWClass *pClass, uint8_t nLevel, uint8_t nSpellLevel, [[maybe_unused]] uint8_t nClass,
+    uint16_t nRace, uint8_t nCastingAbilityBase)
 {
-    uint8_t result = 0;
-    uint8_t spellLevelsPerLevel = 0;
-    uint8_t spellsKnownPerSpellLevel = 0;
-    uint8_t spellsKnownPreviousSpellLevel = 0;
+    if (nLevel < 1 || nLevel > MAX_LEVEL_MAX) return 0;
 
-    if (pClass->m_lstSpellKnownTable)
+    if (!pClass->m_lstSpellKnownTable) return 0;
+
+    int nSpellLevelForLevel = nLevel <= CORE_MAX_LEVEL
+        ? pClass->m_lstSpellLevelsPerLevel[nLevel - 1]
+        : s_nSpellLevelsPerLevelAdded[pClass->m_nName][nLevel - 1];
+
+    if (nSpellLevel >= nSpellLevelForLevel) return 0;
```

Comment:
I think the original has a `nSpellLevel < 0 || ` check here. Without it, it might access the array at an invalid index